### DISCUSSION TITLE
macOS Coreminer Setup Instructions and Guide for Beginners

### DISCUSSION THREAD

**Discussion by tgh90** [catchthatrabbit/coreminer#28] [Operating System]
### macOS Coreminer Setup Instructions and Guide for Beginners

1. Open the LaunchPad, search for Terminal and open it. This is a Command Line Interface (CLI) of your operating system and you control it with commands. Every command I am showing in this guide is one line even though it could happen that it is shown in two lines here on GitHub. First, we need to install a package manager called **Homebrew**:
`/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`
This might take several minutes because it also installs important dependencies like Xcode from Apple. Wait until it is finished.

2. Now we can use **Homebrew** to install the virtual machine **Multipass**: `brew install --cask multipass`
    This also takes some time to get installed. Keep in mind that your operating system might need to be updated.

3. Create a new instance within **Multipass**. The first so-called flag **-n** gives the instance a name. The second flag **-c** sets the number of cores. To check your CPU number of cores click on the top left Apple Logo and click on **About This Mac**. There you can find information about your chip. The third flag **-m** is about memory usage in the instance. I recommend using 1G memory for each core. Change the number of cores and memory usage so that it fits your system:
`multipass launch 22.04 -n coreminer -c 8 -m 8G`

4. You are still on your macOS system. To go into the shell of the Multipass instance type in: `multipass shell coreminer`
    To exit the Multipass instance shell type (don't use it now): `exit`
    The color should change when you are in the shell.

5. Now inside the instance of a Ubuntu 22.04 virtual machine we can start installing the miner software. To see in which directory you are currently in type in `la` or `ls`. If nothing is shown there is no folder yet in the current directory.
Create a new folder: `mkdir coreminer`

6. Change directory: `cd coreminer`

7. Download the Coreminer software. The current version used in this guide v0.19.80.
    For a x86_64 CPU it's:
    `wget -P ./ "https://github.com/catchthatrabbit/coreminer/releases/download/v0.19.80/coreminer-linux-x86_64.tar.gz"`
    For an ARM CPU (M1 Silicon) it's:
    `wget -P ./ "https://github.com/catchthatrabbit/coreminer/releases/download/v0.19.80/coreminer-linux-arm64.tar.gz"`

8. Extract coreminer-linux-x86_64.tar.gz: `tar -xf coreminer-linux-x86_64.tar.gz`
     Or for coreminer-linux-arm64.tar.gz: `tar -xf coreminer-linux-arm64.tar.gz`

9. Change directory: `cd coreapp`

10. Install bc: `sudo apt-get install bc`

11. Start the Coreminer: `bash mine.sh`

For the Core Coin (XCB) Wallet go to: [Wallet Generator](https://github.com/core-coin/wallet-generator)

To go outside of a folder you have to use: `cd ..`

You can stop the miner with 2x **Ctrl+C**. Every time you restart the Multipass Ubuntu 22.04 CLI you just have to open the shell `multipass shell coreminer` and change directory `cd coreminer/coreapp/` and start the Coreminer `bash mine.sh`. Also, you can combine them in one command `cd coreminer/coreapp/ && bash mine.sh`. Sometimes you need to restart the Multipass Ubuntu 22.04 CLI `multipass restart coreminer` (new updates or for a bug fix).

If you want to change the Coreminer settings you have to make changes in the pool.cfg file in the Coreminer folder. You can open it with e.g. `nano pool.cfg`. To save it hit **Ctrl+X**, **Y**, and then **Enter**. Alternatively, you can also remove `rm pool.cfg` the pool.cfg file and start the Coreminer again.

If there is a new version of the Coreminer ([Releases](https://github.com/catchthatrabbit/coreminer/releases)) navigate outside of the **coreminer** folder and remove it with `rm -r coreminer`. Then repeat this guide (step 5) but change the version number on step 7 to the newer version or copy the link and paste it between the quotation marks to replace the download link. Use the newer versions at your own risk because you might need to fix some things to make them work.

You don't have to type in every command twice. Just use the arrow keys up and down on your keyboard to navigate through the history of your previously used commands.

To make a software update for the virtual machine use this command: `sudo apt update && sudo apt upgrade -y`.

**Comment by Kuvala888** [comment]
I changed the num_cores (4) and mem_size (8GB). now iam unable to reduce it cause it gets back to the 4core-8GB setting once I start the mining. why ?

**Reply by tgh90** [reply]
> I changed the num_cores (4) and mem_size (8GB). now iam unable to reduce it cause it gets back to the 4core-8GB setting once I start the mining. why ?

If you follow all the steps I wrote down it should work. At least for me it works. Changed it a couple of times to try different things. If you don't find a solution maybe starting another Multipass instance helps. You can create as many as you want.

**Reply by oldDeathman** [reply]
Hi @tgh90, 

maybe it helps some users to precise step 10 a little bit:
 
9. Load with launchd: launchctl load /Library/LaunchDaemons/com.canonical.multipassd.plist

10. Switch back to the previous terminal (if no new terminal session has been used, just type "exit" to end root user session) and restart Multipass instance: multipass start NAME

**Comment by Kuvala888** [comment]
first of all, thanks a lot for that detailed instructions. miner is running. installed another Multipass instance. somehow it doesn't want to save the change in the settings (number_core, memory_size). I finish the adjustment with Ctrl+X, Y and then Enter. when I start the miner it again is in its default mode of 1core 1gb ram

**Comment by hmercker** [comment]
what are the commands to use to increase the number of cores on ubuntu? also what are the down sides by increasing the number of cores in use?

**Reply by Kuvala888** [reply]
well iam a complete beginner but its written here : 

open Terminal :
Get admin permissions: sudo -s
Enter your password.

You can see your current directory with: ls

Change directory to: cd ../../var/root/Library/Application\ Support/multipassd/qemu/

Open file multipassd-vm-instances.json: nano multipassd-vm-instances.json
(The file can also be in: cd ../../var/root/Library/Application\ Support/multipassd/)

In the file you can change number of cores ("num_cores") and memory size ("mem-size"). Default is 1 core and 1 GB of ram. To save it hit Ctrl+X, Y and then Enter.


increasing the cores will end up in more mining power but more cpu usage (up to 100%) which is long term not healthy for a normal computer

**Reply by hmercker** [reply]
thank you for your help - it seems to work for iOS, I cannot get the same for ubuntu

**Reply by cryptix83** [reply]
Hi, tried to set up the mining but get when i want top start mining in the last step the following error: ./mine.sh: line 97: ./coreminer: cannot execute binary file.

Can anybody help to resolve this ?

**Comment by oldDeathman** [comment]
The multipass machine worked well for about 30 hours on my MacBook Air M1. Then the MacBook was completely frozen. After a restart I was not able to bring the Multipass machine up again. 
Even delete and purge of the machine did not help. 

> user@macbook ~ % multipass shell Coreminer
> start failed: The following errors occurred:                                    
> Coreminer: timed out waiting for response
> user@macbook ~ % multipass list            
> Name                    State             IPv4             Image
> Coreminer              Unknown           --               Ubuntu 22.04 LTS`

I shut down the Mac several times, but it did not help as well. Has anyone any idea how to solve this or what might be the problem? Can I have a look in any log files?
